src/endpoint/sts/sts_rest.js

```diff
@@ -0,0 +1,315 @@
+    } catch (err) {
+        dbg.error('authenticate_request: ERROR', err.stack || err);
+        if (err.code) {
+            throw err;
+        } else {
+            throw new StsError(StsError.AccessDeniedException);
+        }
+    }
+}
+
+async function authorize_request(req) {
+    await Promise.all([
+        req.sts_sdk.authorize_request_account(req),
+        authorize_request_policy(req)
+    ]);
+}
+
+async function authorize_request_policy(req) {
+    if (req.op_name !== 'post_assume_role') return;
+    const account_info = await req.sts_sdk.get_assumed_role(req);
+    const assume_role_policy = account_info.role_config && account_info.role_config.assume_role_policy;
+    if (!assume_role_policy) throw new StsError(StsError.AccessDeniedException);
+    const arn_path = _get_arn_from_req_path(req);
+    const method = _get_method_from_req(req);
+    const auth_token = req.sts_sdk.get_auth_token();
+    let cur_account;
+    if (auth_token && auth_token.access_key) {
+        cur_account = await req.sts_sdk.rpc_client.account.read_account({});
+        // system owner by design can always assume role policy of any account
+        if ((cur_account.email.unwrap() === _get_system_owner().unwrap()) && req.op_name.endsWith('assume_role')) return;
+    }
+    const permission = has_assume_role_permission(assume_role_policy, method, arn_path);
+    if (permission === 'DENY' || permission === 'IMPLICIT_DENY') {
+        throw new StsError(StsError.AccessDeniedException);
+    }
+    // permission is ALLOW, can return without error
+}
+
+function _get_system_owner() {
+    const system_store = require('../../server/system_services/system_store').get_instance(); // eslint-disable-line global-require
+    let system = system_store.data.systems[0];
+    return system.owner.email;
+}
+
+function _get_method_from_req(req) {
+    const sts_op = sts_utils.OP_NAME_TO_ACTION[req.op_name];
+    if (!sts_op) {
+        dbg.error(`Got a not supported STS op ${req.op_name} - doesn't suppose to happen`);
+        throw new StsError(StsError.InternalFailure);
+    }
+    return sts_op;
+}
+
+function _get_arn_from_req_path(req) {
+    const account = req.sts_sdk.requesting_account;
+    let arn_path = account.email.unwrap();
+    return arn_path;
+}
+
+function parse_op_name(req, action_and_params) {
+    const method = req.method.toLowerCase();
+    const action = action_and_params.slice('Action='.length, action_and_params.indexOf('&'));
+    if (ACTIONS[action]) {
+        return `${method}_${ACTIONS[action]}`;
+    }
+    throw new StsError(StsError.NotImplemented);
+}
+
+function parse_params(action_and_params) {
+    const params = {};
+    const params_arr = action_and_params.split('&');
+    params_arr.map(param => {
+        const entry = param.split('=');
+        if (entry[0] === 'Action' || entry[0] === 'Version') return {};
+        params[_.snakeCase(_.lowerCase(entry[0]))] = decodeURIComponent(entry[1]);
```

Comment:
I don't think you mean to use  `_.lowerCase` and then `_.snakeCase`  because by lowering all letters first you don't get the typical word separation when the input is CamelCase 

`CamelCase => (lowercase) camelcase => (snakecase) camelcase`
or
`CamelCase => (snakecase) camel_case`